### PR TITLE
Added Feature - Glue job continuously log printing in Airflow task logs 

### LINKED ISSUE #26196: Amazon Glue job providers not printing log when job get completed or failed.
### Apache Airflow Provider(s)

amazon

### Versions of Apache Airflow Providers

5.0.0

### Apache Airflow version

`>= 2.3.2`

### Operating System

Linux

### Deployment

Docker-Compose

### Deployment details

docker-compose 

### What happened

The method **job_completion** of **GlueJobHook** call **print_job_logs** in **finally** block  that will be never call when job get **completed** either with successfully state or failed state since when job get completed successfully it return value by using **return** statement (that will not execute **finally** block) and similarly in case of **failure** it will **raise** exception that will also not execute the **finally** block.
Due to that airflow does not show Glue job logs from **`CloudWatch`**.

```

def job_completion(self, job_name: str, run_id: str, verbose: bool = False) -> Dict[str, str]:
        """
        Waits until Glue job with job_name completes or
        fails and return final state if finished.
        Raises AirflowException when the job failed
        :param job_name: unique job name per AWS account
        :param run_id: The job-run ID of the predecessor job run
        :param verbose: If True, more Glue Job Run logs show in the Airflow Task Logs.  (default: False)
        :return: Dict of JobRunState and JobRunId
        """
        failed_states = ['FAILED', 'TIMEOUT']
        finished_states = ['SUCCEEDED', 'STOPPED']
        next_log_token = None
        job_failed = False

        while True:
            try:
                job_run_state = self.get_job_state(job_name, run_id)
                if job_run_state in finished_states:
                    self.log.info('Exiting Job %s Run State: %s', run_id, job_run_state)
                    return {'JobRunState': job_run_state, 'JobRunId': run_id}
                if job_run_state in failed_states:
                    job_failed = True
                    job_error_message = f'Exiting Job {run_id} Run State: {job_run_state}'
                    self.log.info(job_error_message)
                    raise AirflowException(job_error_message)
                else:
                    self.log.info(
                        'Polling for AWS Glue Job %s current run state with status %s',
                        job_name,
                        job_run_state,
                    )
                    time.sleep(self.JOB_POLL_INTERVAL)
            finally:
                if verbose:
                    next_log_token = self.print_job_logs(
                        job_name=job_name,
                        run_id=run_id,
                        job_failed=job_failed,
                        next_token=next_log_token,
                    )
```

### What you think should happen instead

It should print the log in all cases, (failure or success) when **`verbose=True`.**

### How to reproduce

Use latest version of **`amazon providers==5.0.0`** and create a Airflow task for any Glue job.
Make sure to pass **`verbose = True`** in **`GlueJobOperator`** as below:

```
job_task = GlueJobOperator(task_id ="testJob", 
                                               job_name= "<Glue job name>",
                                               job_desc="Test Job",
                                               region_name="ap-south-1",
                                               verbose=True,
                                               script_location="s3://..//../", 
                                               num_of_dpus=2) 
```

### Anything else

This is a code bug that will result no `**cloudwatch**` log print in every case. I will provide the solution with enhancement for continuous logging.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### LINKED ISSUE #26203: Continuously logging and Batch logging mode for the aws Glue job's execution logs
### Description

Currently Airflow does not print the **`Glue job`** execution logs due to the bug highlighted in **Related Issue**. This feature will enable the **Continuously logging** and **Batch logging**  of **`Glue jobs`** in **`Airflow`**  based on the **`GlueJobOperator`** input parameters. 

### Use case/motivation

I have orchestrated multiple Glue jobs with Airflow, Due to existing code  bug and no option for continuously logging, I have to check **`AWS Cloudwatch`** explicitly for Glue job's logs. After this feature, we will not require to check the logs at  **`AWS Cloudwatch`**, **the same logs will be available at **`Airflow`** for monitoring the Glue jobs.** 

### Related issues

https://github.com/apache/airflow/issues/26196

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### LINKED ISSUE #26511: Retrieve AWS Cloudwatch logs using aws_conn_id connection for GlueOperator
### Apache Airflow version

2.4.0

### What happened

We are using cross account AWS connection to invoke a Glue job using GlueOperator. When the verbose flag is set we expected the cloudwatch logs to be retrieved from the AWS account where glue job was executed and displayed in Airflow logs.
We constantly receive below error on Airflow logs rather than getting cloudwatch logs

```
Polling for AWS Glue Job <Job name> current run state with status RUNNING
No new Glue driver logs found. This might be because there are no new logs, or might be an error.
If the error persists, check the CloudWatch dashboard at: https://ap-southeast-1.console.aws.amazon.com/cloudwatch/home
```

**Underlying issue**
Below code suggest that it creates a new connection to retrieve logs on same account as the invoker of glue job.

[airflow/glue.py at cc4f245758340f5fc278bbbdc958a40b85f39bb8 · apache/airflow](https://github.com/apache/airflow/blob/cc4f245758340f5fc278bbbdc958a40b85f39bb8/airflow/providers/amazon/aws/hooks/glue.py#L155) 

But as we are executing the glue job from a different account; the code needs updates to share the same connection in retrieving the logs. 

### What you think should happen instead

When AWS GlueOperator is called with a aws_conn_id parameter and verbose flag we expect

1. Glue job to be executed in the aws account as per aws_conn_id
2. Cloudwatch logs to be relieved from aws account as per aws_conn_id and displayed in Airflow logs

**Proposed Solution**

1. Dont create a new boto3 client for logs (https://github.com/apache/airflow/blob/cc4f245758340f5fc278bbbdc958a40b85f39bb8/airflow/providers/amazon/aws/hooks/glue.py#L155)
2. Create a client using a derived method of base AwsBaseHook classes method get_client_type (https://github.com/apache/airflow/blob/4bf0cb98724a2cf04aab6359881a87aeb9cec0ce/airflow/providers/amazon/aws/hooks/base_aws.py#L438)
3. Update glue Hooks print_job_logs method to use boto3 get_log_events function rather than filter_log_events function (https://github.com/apache/airflow/blob/4bf0cb98724a2cf04aab6359881a87aeb9cec0ce/airflow/providers/amazon/aws/hooks/glue.py#L165). This has been identified as another problem where the filter operation is unsuitable for retrieving verbose logs. When filter_log_events is used with no filter the output doesn't have events.

### How to reproduce

1. submit a glue job in a different AWS account using GlueOperator as below

```
with DAG(
    dag_id="sample_dag",
    description="Sample DAG testing",
    schedule_interval=None,
    start_date=datetime(2022, 9, 1),
    catchup=False,
) as dag:
    submit_glue_job = SeekGlueJobOperator(
        aws_conn_id="different_aws_account",
        task_id="submit_glue_job",
        job_name=job_name,
        wait_for_completion=True,
        retry_limit=1,
        script_location=None,
        iam_role_name="iam-role-for-glue-job-invocation",
        script_args={},
        run_job_kwargs={"NumberOfWorkers": 2, 'WorkerType': "Standard"},
        verbose=True,
        region_name="ap-southeast-1"
    )
```

4. Check Airflow logs to see if cloudwatch logs are retrieved and displayed

### Operating System

Mac 12.6

### Versions of Apache Airflow Providers

_No response_

### Deployment

Docker-Compose

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by nikhi-suthar**
Added Continuously logging feature for Glue jobs and fixed  existing bug for not printing Cloduwatch logs.

Following are the changes done:

- **GlueJobOperator**
     - Added new  input parameters name as **`continuous_logging`** for Continuous logging. Default value is False

- **GlueJobHooks**
    - Added new  input parameters name as **`continuous_logging`** for Continuous logging in the method **`job_completion`**. 
    - Added the logic to print available logs from **Cloudwatch** to Airflow logs.
    - Added new method **`print_output_logs`** for handling continuously logging. 
    - Fixed the existing bug for not executing method **`print_jobs_logs`**
    - Change the method for printing selective logs from  filter mechanism to **get log event** method.
    - Added Condition for **Batch** , **Continue** and **no logging** in **`job_completion`** method

Closes: #26203
Closes: https://github.com/apache/airflow/issues/26196

### CODE REVIEW

File: airflow/providers/amazon/aws/hooks/glue.py [hunk 1]
```diff
@@ -151,85 +153,150 @@ def print_job_logs(
...
-
-        filter_pattern = FAILURE_LOG_FILTER if job_failed else DEFAULT_LOG_FILTER
+        """Prints the batch of Glue cloudwatch logs to the Airflow task log and returns nextToken."""
+        credentials = self.get_credentials(region_name=self.conn_region_name)
+        log_client = boto3.client(
+            'logs',
+            region_name=self.conn_region_name,
+            aws_access_key_id=credentials.access_key,
+            aws_secret_access_key=credentials.secret_key,
+        )
```
Comment (Taragolis) [inline airflow/providers/amazon/aws/hooks/glue.py]: Might be rather than directly use `boto3` use AwsBaseHook.get_session method instead, so it could be replaced by

```python
log_client = self.get_session(region_name=self.conn_region_name).client('logs')
```

File: airflow/providers/amazon/aws/hooks/glue.py [hunk 2]
```diff
@@ -151,85 +153,150 @@ def print_job_logs(
...
+        self.log.info("region name - %s", region_name)
+        credentials = self.get_credentials(region_name=region_name)
         failed_states = ['FAILED', 'TIMEOUT']
         finished_states = ['SUCCEEDED', 'STOPPED']
+        log_client = boto3.client(
+            'logs',
+            region_name=region_name,
+            aws_access_key_id=credentials.access_key,
+            aws_secret_access_key=credentials.secret_key,
+        )
```
Comment (Taragolis) [inline airflow/providers/amazon/aws/hooks/glue.py]: Same as above

File: airflow/providers/amazon/aws/hooks/glue.py (hunk 1 above)
Comment (Taragolis) [reply airflow/providers/amazon/aws/hooks/glue.py]: BTW, in your case you do not provide `aws_session_token` into the client. If user follow aws best practices and use temporary credentials obtain by role_arn, EC2 Instance IAM role, ECS Task IAM Role, IRSA EKS and etc. then token should be provided otherwise error will raised.

I thought token not required only for permanent credentials

File: airflow/providers/amazon/aws/hooks/glue.py [hunk 3]
```diff
@@ -151,85 +153,150 @@ def print_job_logs(
...
-                )
-            else:
-                response = log_client.filter_log_events(
-                    logGroupName=log_group_name,
-                    logStreamNames=[run_id],
-                    filterPattern=filter_pattern,
-                )
-            if len(response['events']):
+            self.log.info('Glue Job Run Logs')
+            response = log_client.get_log_events(logGroupName=log_group_name, logStreamName=run_id)
```
Comment (ferruzzi) [inline airflow/providers/amazon/aws/hooks/glue.py]: Why the change from filter_log_events to get_log_events?   If you're going to do that, then also drop the filter constants on L33-34

File: airflow/providers/amazon/aws/hooks/glue.py [hunk 4]
```diff
@@ -151,85 +153,150 @@ def print_job_logs(
...
-                        'Polling for AWS Glue Job %s current run state with status %s',
-                        job_name,
-                        job_run_state,
-                    )
-                    time.sleep(self.JOB_POLL_INTERVAL)
-            finally:
-                if verbose:
-                    next_log_token = self.print_job_logs(
+                job_run = glue_client.get_job_run(JobName=job_name, RunId=run_id, PredecessorsIncluded=True)
+                job_run_state = job_run['JobRun']['JobRunState']
```
Comment (ferruzzi) [inline airflow/providers/amazon/aws/hooks/glue.py:253]: Any reason not to keep using the existing get_job_state helper here?

File: airflow/providers/amazon/aws/hooks/glue.py [hunk 5]
```diff
@@ -151,85 +153,150 @@ def print_job_logs(
...
+                            run_id=run_id,
+                            log_group_name=log_group_name,
+                            log_client=log_client,
+                            next_token=next_log_token,
+                        )
+                elif mode == 'BATCH':
+                    msg = f"The Job {job_name} with run id {run_id} is : {job_run_state}"
+                    self.log.info(msg)
+                else:
+                    self.log.info(msg)
```
Comment (ferruzzi) [inline airflow/providers/amazon/aws/hooks/glue.py:233]: I know it's silly, but can you humour me and add a blank line here between the `if` blocks to visually separate them?
```suggestion
                    self.log.info(msg)
                    
```

File: airflow/providers/amazon/aws/hooks/glue.py [hunk 6]
```diff
@@ -151,85 +153,150 @@ def print_job_logs(
...
+                        next_log_token = token
+                        token = self.print_output_logs(
+                            job_name=job_name,
+                            run_id=run_id,
+                            log_group_name=log_group_name,
+                            log_client=log_client,
+                            next_token=next_log_token,
+                        )
+                elif mode == 'BATCH':
+                    msg = f"The Job {job_name} with run id {run_id} is : {job_run_state}"
```
Comment (ferruzzi) [inline airflow/providers/amazon/aws/hooks/glue.py]: Is this not basically the same value as it is already assigned on L254?   Do we need this assignment here?

**Comment by potiuk** [comment]
Needs conflict resolving and answering @ferruzzi 's comments.

File: airflow/providers/amazon/aws/hooks/glue.py (hunk 1 above)
Comment (nikhi-suthar) [reply airflow/providers/amazon/aws/hooks/glue.py:163]: > 

I tried that but the issue is, in Glue connector, we are already using one client for Glue from AWSBaseHook object, so for cloudwatch I really do not want to create one more object of AWSBaseHook t get the boto3 client that's a reason I have added it directly through boto3. Correct

File: airflow/providers/amazon/aws/hooks/glue.py (hunk 3 above)
Comment (nikhi-suthar) [reply airflow/providers/amazon/aws/hooks/glue.py]: filter_log_events will not give logs and also in the case of a custom log, it will not be useful. get_logs_events I have tested and it is working fine.  Agree with you regarding extra filter constants, I will drop and push new changes.

File: airflow/providers/amazon/aws/hooks/glue.py (hunk 1 above)
Comment (nikhi-suthar) [reply airflow/providers/amazon/aws/hooks/glue.py]: In the existing Glue hook, we are using boto3 directly for log_client. I guess the reason for that will be a conflict of kwargs['client_type"]. since we are already using a glue client from AWSBaseHook in the same code, creating a new client for logs from the same AWSBaseHook class is not possible. I agree with your point that it might cause issue with role-based access.. let me check how we can solve it. I guess I should declare AWSLogsHook object and get a session from that only. I will try that and push new changes soon. Meanwhile, if you have any other workaround please feel free to provide the same.

File: airflow/providers/amazon/aws/hooks/glue.py (hunk 5 above)
Comment (nikhi-suthar) [reply airflow/providers/amazon/aws/hooks/glue.py:233]: Sure will do

File: airflow/providers/amazon/aws/hooks/glue.py (hunk 6 above)
Comment (nikhi-suthar) [reply airflow/providers/amazon/aws/hooks/glue.py]: I had added other use cases, but currently it is not required. Will remove it.

File: airflow/providers/amazon/aws/hooks/glue.py (hunk 1 above)
Comment (Taragolis) [reply airflow/providers/amazon/aws/hooks/glue.py]: > creating a new client for logs from the same AWSBaseHook class is not possible

Yes you can not create different client type by call `AWSBaseHook.get_client_type` since latest changes.

https://github.com/apache/airflow/blob/3b6176929ccc0351554e5331d920c4ec65b33a38/airflow/providers/amazon/aws/hooks/base_aws.py#L438-L454

But you still can create `boto.session.Session()` and can use this session for create whatever client you wanted

https://github.com/apache/airflow/blob/3b6176929ccc0351554e5331d920c4ec65b33a38/airflow/providers/amazon/aws/hooks/base_aws.py#L432-L436

**Generic approach for boto3 client**
```python
session = boto3.session.Session(...)
client = session.client("awesome-client", ...)
```

**How it could be archived by `AwsGenericHook` / `AWSBaseHook` and any their child**
```python
hook = GlueJobHook(aws_conn_id="awesome_aws_conn_id", ...)
session = self.get_session(...)
client = session.client("awesome-client", ...)
```

Comment (nikhi-suthar) [reply airflow/providers/amazon/aws/hooks/glue.py]: I have taken care of Log client from AWSBaseHook. It is working fine now. Please review now and let me know any further changes require.

**Comment by ferruzzi** [comment]
Changes and replies look good to me.  Thanks for those, and for your patience. :+1:

File: airflow/providers/amazon/aws/hooks/glue.py [hunk 7]
```diff
@@ -202,34 +204,52 @@ def job_completion(self, job_name: str, run_id: str, verbose: bool = False) -> D
         failed_states = ['FAILED', 'TIMEOUT']
         finished_states = ['SUCCEEDED', 'STOPPED']
         next_log_token = None
-        job_failed = False
+        glue_client = self.get_conn()
+        cw_logs = AwsLogsHook()
```
Comment (Taragolis) [inline airflow/providers/amazon/aws/hooks/glue.py]: Without any parameters to the hook it will use default boto3 strategy, and `cw_logs.get_conn()` turned into `boto3.client("logs")`.

```suggestion
        cw_logs = AwsLogsHook(aws_conn_id=self.aws_conn_id, region_name=self.region_name)
```

But it really nice that you move code related to AWS CloudWatch to the `AwsLogsHook`

Comment (nikhi-suthar) [reply airflow/providers/amazon/aws/hooks/glue.py]: Thanks, @Taragolis  I have made changes as per the suggestion.

**Comment by potiuk** [comment]
For some reason the future.__annotations__ import is duplicated ?

File: airflow/providers/amazon/aws/hooks/glue.py [hunk 8]
```diff
@@ -17,12 +17,15 @@
 # under the License.
 from __future__ import annotations
 
+from __future__ import annotations
+
```
Comment (Taragolis) [inline airflow/providers/amazon/aws/hooks/glue.py]: @nikhi-suthar this one

Comment (nikhi-suthar) [reply airflow/providers/amazon/aws/hooks/glue.py]: Changes Done

**Comment by o-nikolas** [comment]
@nikhi-suthar 
Are you still attending this PR? It looks like the glue tests are timing out after your changes.

**Comment by shubham22** [comment]
@nikhi-suthar - checking once more, are you still working on this? cc: @o-nikolas

**Comment by Dearkano** [comment]
Hi is there any update on this PR, thanks!

**Comment by potiuk** [comment]
> Hi is there any update on this PR, thanks!

Not really - as you see it's closed. But if you want to pick it up where the original author left it - feel free to continue in a new PR.
